controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -144,6 +144,66 @@ public void getCurrentSegmentsKeyValueTable(KeyValueTableInfo request, StreamObs
                 responseObserver);
     }
 
+    @Override
+    public void listKeyValueTablesInScope(Controller.KVTablesInScopeRequest request, StreamObserver<Controller.KVTablesInScopeResponse> responseObserver) {
+
+        String scopeName = request.getScope().getScope();
+        RequestTag requestTag = requestTracker.initializeAndTrackRequestTag(requestIdGenerator.get(),
+                "listKeyValueTables", scopeName);
+        log.info(requestTag.getRequestId(), "listKeyValueTables called for scope {}.", scopeName);
+
+        final AuthContext ctx;
+        if (this.grpcAuthHelper.isAuthEnabled()) {
+            ctx = AuthContext.current();
+        } else {
+            ctx = null;
+        }
+
+        authenticateExecuteAndProcessResults(
+                () -> {
+                    String result = this.grpcAuthHelper.checkAuthorization(
+                            AuthResourceRepresentation.ofScope(scopeName),
+                            AuthHandler.Permissions.READ,
+                            ctx);
+                    log.debug("Result of authorization for [{}] and READ permission is: [{}]",
+                            AuthResourceRepresentation.ofScope(scopeName), result);
+                    return result;
+                },
+                delegationToken -> controllerService
+                        .listKeyValueTables(scopeName, request.getContinuationToken().getToken(), listStreamsInScopeLimit)
+                        .handle((response, ex) -> {
+                            if (ex != null) {
+                                if (Exceptions.unwrap(ex) instanceof StoreException.DataNotFoundException) {
+                                    return Controller.KVTablesInScopeResponse.newBuilder().setStatus(Controller.KVTablesInScopeResponse.Status.SCOPE_NOT_FOUND).build();
+                                } else {
+                                    throw new CompletionException(ex);
+                                }
+                            } else {
+                                log.debug("All KeyValueTables in scope with continuation token: {}", response);
+                                List<KeyValueTableInfo> kvtables = response
```

Comment:
There is one problem with the logic here. And i realize this same problem is present for streams too. 
If we filter out unauthorized tables, then the response we send back may endup being empty while there may still be more records to be read from the continuation token.

So any such filtering should invoke the call with `filtered-out` elements and keep doing it until we have collected `limit` number of elements or have nothing to read anymore.

Reply:
Fixed this.